From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <novau233@163.com>
Date: Sun, 26 May 2024 11:13:14 +0000
Subject: [PATCH] Fix build and set serverName to LightingLuminol


diff --git a/build.gradle.kts b/build.gradle.kts
index ab07c307bea0d3b79a6239bdcca9714e1ddf3b28..82a65f08105a5b7fd4fb0b780259b3a739f43270 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -13,7 +13,7 @@ configurations.named(log4jPlugins.compileClasspathConfigurationName) {
 val alsoShade: Configuration by configurations.creating
 
 dependencies {
-    implementation(project(":luminol-api")) // Folia // Luminol
+    implementation(project(":lightingluminol-api"))//Luminol //LightingLuminol
     implementation("com.electronwill.night-config:toml:3.6.6") // Luminol - Night config
     // Paper start
     implementation("org.jline:jline-terminal-jansi:3.21.0")
diff --git a/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java b/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
index 673a81734a0903a0790f3007772642d3d9a044cc..31f3c9039bf7e8191f1a207276f73cb75dd8f57e 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
@@ -31,7 +31,7 @@ public record ServerBuildInfoImpl(
     private static final String ATTRIBUTE_GIT_COMMIT = "Git-Commit";
 
     private static final String BRAND_PAPER_NAME = "Paper";
-    private static final String BRAND_LUMINOL_NAME = "Luminol";
+    private static final String BRAND_LUMINOL_NAME = "LightingLuminol"; //LightingLuminol
 
     private static final String BUILD_DEV = "DEV";
 
